OPENNLP-1448 - Introduce SLF4J in OpenNLP

rzo1 opened the pull request:
Thank you for contributing to Apache OpenNLP.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with OPENNLP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn clean install at the root opennlp folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file in opennlp folder?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found in opennlp folder?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:

This PR is the first sub-task of [OPENNLP-1447](https://issues.apache.org/jira/browse/OPENNLP-1447).

- It introduces `slf4j-api` in OpenNLP and replaces calls to `System.err` or `System.out` by related logger calls.
- It adds `log4j2` as `runtime` dependency to the OpenNLP distribution + related configuration and adjusts the scripts accordingly

#### Notes for reviewers:

- Revision of log levels is tracked with [OPENNLP-1449](https://issues.apache.org/jira/browse/OPENNLP-1449) and is not scope of the given PR. However, comments on specific log levels are highly welcome, so they can be tracked and incorporated in OPENNLP-1449.
- Fully use of formatting using the possibilities of SLF4J eg.`{}`-replacement is tracked with [OPENNLP-1450](https://issues.apache.org/jira/browse/OPENNLP-1450) and is not scope of the given PR.
- Log output reduction during the build is currently tracked with  [OPENNLP-1451](https://issues.apache.org/jira/browse/OPENNLP-1451). At the moment, I introduced `slf4j-simple` as a `test` dependency and disabled log output. We can discuss how we want to deal with it within OPENNLP-1451.
- If some removed `System.out / System.err` statements (in the tests) are really needed, please leave a comment. We can easily move them to log statements

#### TODOs for reviewers:

- It would be great, if our heavy CLI users like @atarora can provide valuable feedback.
- Due to the absence of a Windows machine, I cannot test the `.bat` changes - maybe some Windows user can have a look here.
- It would be great, if someone with the `eval` datasets available on his/her machine can run the `eval` tests and provide feedback, if needed.
- I disabled a few tests, which relied on overriding `System.out` and gathering the output. Happy to receive ideas on how we can re-write them to work with `slf4j-simple` in the test context


![image](https://user-images.githubusercontent.com/13417392/213663341-0e8a76fa-188b-4137-90e5-4c3f9427a303.png)

rzo1 commented:
Switched it to `DRAFT`, so it cannot be merged accidently but the review would be very much appreciated!

rzo1 commented on opennlp-dl/src/test/java/opennlp/dl/doccat/DocumentCategorizerDLEval.java:
```diff
@@ -29,13 +29,17 @@
 
 public class DocumentCategorizerDLEval extends AbstactDLTest {
 
+  private static Logger logger = LoggerFactory.getLogger(DocumentCategorizerDLEval.class);
```
missing `final`

rzo1 commented on opennlp-dl/src/main/java/opennlp/dl/doccat/DocumentCategorizerDL.java:
```diff
@@ -51,6 +53,7 @@
  */
 public class DocumentCategorizerDL implements DocumentCategorizer {
 
+  private Logger logger = LoggerFactory.getLogger(DocumentCategorizerDL.class);
```
missing `static final`

jzonthemtn commented:
This looks awesome! Thanks for the level of detail on everything. I can volunteer to run the eval tests.

rzo1 commented:
> This looks awesome! Thanks for the level of detail on everything. I can volunteer to run the eval tests.

Sure. Please give it a try.

kinow commented on opennlp-tools/src/test/java/opennlp/tools/cmdline/tokenizer/TokenizerTrainerToolTest.java:
```diff
@@ -104,7 +108,10 @@ public void testTestRunHappyCase() throws IOException {
+  //TODO OPENNLP-1447
+  @Disabled(value = "OPENNLP-1447: These kind of tests won't work anymore. " +
+          "We need to find a way to redirect log output (i.e. implement " +
+          "a custom log adapter and plug it in, if we want to do such tests.")
```
:+1: that's common in Python, but I have never had to do that in Java. Maybe something like: https://stackoverflow.com/questions/29076981/how-to-intercept-slf4j-with-logback-logging-via-a-junit-test (will comment in that issue).

rzo1 replied:
The log watcher approach as outlined in the thread looks promising. I will add a reference in OPENNLP-1447

rzo1 replied:
Already documented by @kinow  -> thanks!

kinow commented on opennlp-tools/src/main/java/opennlp/tools/log/LogPrintStream.java:
```diff
@@ -0,0 +1,87 @@
+import opennlp.tools.commons.Internal;
+
+/**
+ * This class serves as an adapter for a {@link Logger} used withing a {@link PrintStream}.
```
s/withing/within

kinow commented on opennlp-tools/src/main/java/opennlp/tools/log/LogPrintStream.java:
```diff
@@ -0,0 +1,87 @@
+        logger.warn(msg);
+      case ERROR:
+        logger.error(msg);
+    }
```
Isn't there a `logger.log(msg, level)` method in SLF4J? If so this method could be deleted, and the caller code adapted to call that instead? (I think I used a method like that, but can't recall the logging library, nor the programming language :rofl: )

rzo1 replied:
Looked into the interface but it seems, that slf4j doesn't have it

kinow replied:
#TodayILearned 

@rzo1 I think this switch is still missing the `break` ? :car: 

```java
public static void main(String[] args) {
        int level = 0;
        switch (level) {
            case 0:
                System.out.print("TRACE ");
            case 1:
                System.out.print("WARN ");
            case 2:
                System.out.print("INFO ");
            case 3:
                System.out.print("DEBUG");
        }
    }
```

Prints `TRACE WARN INFO DEBUG`, so I think in the end this would call all the `logger.*` methods?

rzo1 replied:
Yes - you are right ;-) - will fix

kinow reviewed (commented):
Reviewed 63 out of 150 files :smiling_face_with_tear: Next time I have time I can either finish it or try running the code locally to verify it's working and its output. Thanks a lot!

jzonthemtn commented:
I cloned the branch and ran the build. The output from the build is much, much cleaner so that's awesome. The eval tests are also much cleaner and easier to read. It will be nice for applications that use OpenNLP to not be flooded with print statements.

- The eval tests are running and I'll update then they're finished.
- I can also try out the bat files on Windows.

I don't have any significant issues with this. I think this is a great addition and we can always improve it down the road when necessary. We can look at how we can improve those tests that rely on the sys out.

mawiesne reviewed (changes requested):
Thanks @rzo1 for those important changes and the work you put in! Great progress! 

I have mostly minor feedback, but found one or two changes that we need to work out or find consensus on. No hard blockers, though. Needs @jzonthemtn and/or feedback by @kinow et al.

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/parser/Parse.java:
```diff
@@ -319,32 +323,25 @@ public void insert(final Parse constituent) {
       int pn = parts.size();
       for (; pi < pn; pi++) {
         Parse subPart = parts.get(pi);
-        //System.err.println("Parse.insert:con="+constituent+" sp["+pi+"] "+subPart+" "+subPart.getType());
```
Maybe we should re-add these statements and replace them with `trace(...)`. If we ever need to debug it, we can switch to trace to see what happens.

kinow replied:
+1 too. Especially to `trace`. Before since the output of tests & tools was all in `out`/`err` sometimes I would just ignore it as it was too much to parse it (which could possibly hide some errors). Having some like this in a separate level is extremely helpful :pray: :bow: Thanks!

rzo1 replied:
Will change to trace :-)

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/perceptron/SimplePerceptronSequenceTrainer.java:
```diff
@@ -372,14 +362,11 @@ public void nextIteration(int iteration) throws IOException {
-            //System.err.println("updates["+pi+"]["+oi+"]=("+updates[pi][oi][ITER]+","
-            // +updates[pi][oi][EVENT]+","+updates[pi][oi][VALUE]+") + ("+iterations+","+0+","
-            // +params[pi].getParameters()[oi]+") -> "+averageParams[pi].getParameters()[oi]);
           }
```
Trace? Similar to @mawiesne comments below?

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/naivebayes/NaiveBayesTrainer.java:
```diff
@@ -228,7 +233,7 @@ private double trainingStats(EvalParameters evalParams) {
     }
     double trainingAccuracy = (double) numCorrect / numEvents;
-    display("Stats: (" + numCorrect + "/" + numEvents + ") " + trainingAccuracy + "\n");
+    logger.info("Stats: (" + numCorrect + "/" + numEvents + ") " + trainingAccuracy);
```
Variable replacement is tracked with https://issues.apache.org/jira/browse/OPENNLP-1450 (next iteration, next topic :-) )

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/perceptron/PerceptronTrainer.java:
```diff
@@ -439,22 +442,10 @@ private double trainingStats(EvalParameters evalParams) {
     }
     double trainingAccuracy = (double) numCorrect / numEvents;
-    display("Stats: (" + numCorrect + "/" + numEvents + ") " + trainingAccuracy + "\n");
+    logger.info("Stats: (" + numCorrect + "/" + numEvents + ") " + trainingAccuracy);
```
https://issues.apache.org/jira/browse/OPENNLP-1450

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/perceptron/SimplePerceptronSequenceTrainer.java:
```diff
@@ -235,14 +240,8 @@ public AbstractModel trainModel(int iterations, SequenceStream<Event> sequenceSt
 
   private void findParameters(int iterations) throws IOException {
-    display("Performing " + iterations + " iterations.\n");
+    logger.info("Performing " + iterations + " iterations.\n");
```
https://issues.apache.org/jira/browse/OPENNLP-1450

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/parser/treeinsert/Parser.java:
```diff
@@ -444,33 +450,33 @@ else if (1 - cprobs[completeIndex] > probMass) { //just incomplete advances
-                  if (debugOn)
-                    System.out.println("Advancing both complete and incomplete nodes; c="
+                  if (logger.isDebugEnabled())
+                    logger.debug("Advancing both complete and incomplete nodes; c="
```
https://issues.apache.org/jira/browse/OPENNLP-1450

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/formats/masc/MascDocument.java:
```diff
@@ -353,7 +353,7 @@ private void addPennTags(Map<String, Map<Integer, ?>> tagMaps) throws IOExceptio
       for (Map.Entry<Integer, int[]> token : tokenToQuarks.entrySet()) {
         if (token.getValue().length == 0) {
-          System.err.println("[ERROR] Token without quarks: " + token.getKey());
+          logger.error("Token without quarks: {}", token.getKey());
```
I agree with `warn` here. We can change within https://issues.apache.org/jira/browse/OPENNLP-1449 (next sub-task, next topic).

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/maxent/quasinewton/ParallelNegLogLikelihood.java:
```diff
@@ -149,7 +154,7 @@ private void computeInParallel(double[] x, Class<? extends ComputeTask> taskClas
 
     } catch (Exception e) {
-      e.printStackTrace();
+      logger.warn(e.getLocalizedMessage(), e);
```
+1 -> https://issues.apache.org/jira/browse/OPENNLP-1449

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/tokenize/lang/en/TokenSampleStream.java:
```diff
@@ -118,7 +121,7 @@ public void remove() {
   private static void usage() {
-    System.err.println("TokenSampleStream [-spans] < in");
-    System.err.println("Where in is a space delimited list of tokens.");
+    logger.error("TokenSampleStream [-spans] < in");
```
Yes. I guess, that `err` was used here for a nice "red" console message. We should change it to `info` with https://issues.apache.org/jira/browse/OPENNLP-1449

rzo1 commented on opennlp-brat-annotator/src/main/java/opennlp/bratann/NameFinderAnnService.java:
```diff
@@ -41,15 +43,16 @@
     if (args.length == 0) {
-      System.out.println("Usage:");
-      System.out.println("[NameFinderAnnService -serverPort port] [-tokenizerModel file] "
+      logger.info("Usage:");
```
It basically depends on the logger impl behind the slf4j facade. We might miss some "beautiful" indentation for progress reporting, but overall I don't think, that we can "miss" a message depending on the backing logger impl. used. The good thing is, that it allows our users to configure additional file appenders, so a training session is recorded.

rzo1 commented:
> Thanks @rzo1 for those important changes and the work you put in! Great progress!
> 
> I have mostly minor feedback, but found one or two changes that we need to work out or find consensus on. No hard blockers, though. Needs @jzonthemtn and/or feedback by @kinow et al.

Thanks for the review. We discussed the importance of the sysouts with a :coffee: a while ago. So I am definitly in favour for **not** removing them. We should maybe replace them with trace and actually comment them in, so we have it in place if needed. Don't think we need to have it as dead code.

Regarding the variable substituion / replacement / injection: Will revice and change with https://issues.apache.org/jira/browse/OPENNLP-1450

Regarding log levels: Will revise and change with
https://issues.apache.org/jira/browse/OPENNLP-1449

I will resolve the current merge conflict now.

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/maxent/GISTrainer.java:
```diff
@@ -407,9 +399,9 @@ public GISModel trainModel(int iterations, DataIndexer di, Prior modelPrior, int
-    display("\tNumber of Event Tokens: " + numUniqueEvents + "\n");
-    display("\t    Number of Outcomes: " + numOutcomes + "\n");
-    display("\t  Number of Predicates: " + numPreds + "\n");
+    logger.info("Number of Event Tokens: {}", numUniqueEvents);
```
I would consider it a group, which could be logged in a one-liner but happy to hear from @kinow and @jzonthemtn

mawiesne replied:
+1 for: one-liner/group, maybe formatted as before with `\t` indentation.

kinow replied:
+1 for either one-liner/group too, as long as the information is there in the logs as appears as a single block without risks of other threads logging anything :+1:

rzo1 replied:
I will change it to a one-liner with  \t formatting as before.

kinow reviewed (commented):
> 150 / 150 files viewed

:tada: finished reviewing it! It must have been even more challenging to write these changes. Thanks a lot @rzo1 . See one comment about a switch, and there are pending discussions about groups in log output, and about re-adding some log output that was removed now in `trace` or at another level.

Thanks heaps!

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/perceptron/PerceptronTrainer.java:
```diff
@@ -265,25 +270,25 @@ public AbstractModel trainModel(int iterations, DataIndexer di, int cutoff, bool
-    display("\tNumber of Event Tokens: " + numUniqueEvents + "\n");
-    display("\t    Number of Outcomes: " + numOutcomes + "\n");
-    display("\t  Number of Predicates: " + numPreds + "\n");
+    logger.info("Number of Event Tokens: {}", numUniqueEvents);
```
done

rzo1 commented on opennlp-tools/src/main/java/opennlp/tools/ml/perceptron/SimplePerceptronSequenceTrainer.java:
```diff
@@ -195,11 +199,12 @@ public AbstractModel trainModel(int iterations, SequenceStream<Event> sequenceSt
-    display("done.\n");
+    logger.info("done.");
+
+    logger.info("Number of Event Tokens: {}", numEvents);
```
done

rzo1 commented on opennlp-distr/src/main/assembly/bin.xml:
```diff
@@ -48,7 +48,17 @@
+				<include>log4j2.xml</include>
+			</includes>
+		</fileSet>
+
```
Fixed via autoformatting of whole file! Thanks

rzo1 commented:
I just re-added `System.out` / `System.err` as `trace` (as discussed), switched to grouped 1-liner with `\t`. Fixed whitespace / tab issue ìn `bin.xml` 

I will remove the DRAFT label now and wait for eval feedback by @jzonthemtn

kinow reviewed (approved):
Bravo :clap:

jzonthemtn commented:
Sorry, @rzo1, I was slow. No issues with the tests. I think the Windows changes are fine but we can always address anything in a different Jira ticket if something comes up. Thanks for merging this - I'm really excited about this one.
